### ISSUE TITLE
Calibration with non-uniform widths at test time

### ISSUE THREAD

**Issue by MohammadSamragh** [JiahuiYu/slimmable_networks#27]
Hi Jiahui,

Thannk you for this great work! I was trying to calibrate the BN layers after loading it in test time but I failed. Can you please help me through this? I did the following steps:

1- First, I changed the USConv, USBatcnNorm, and USLinear code such that they can have different with_multiplier for the input and output channels. This way, it is not required that each layer has the same with_multiplier. From there, I was able to run a model with non-uniform width_multipliers at each layer. For instance, I can have [L1: 0.3, L2: 0.75, L3: 0.9, ...]. 

2- The above code runs with no errors. When I set all width_multipliers to the same value (uniformly), it works well and the accuracy is OK. The accuracy, however, is not high for non-uniform widths. For example, [L1: 0.3, L2: 0.3, L3: 0.3, ...] gives me a fare accuracy since all layers have a width of 0.3, but [L1: 0.3, L2: 0.4, L3: 0.7, ...] has a very bad accuracy because the width_multiplier is different across layers.

3- The above problem was inevitable since the batchnorm layers are only calibrated for uniform width_multipliers. I tried to calibrate the batchnorms myself but I could not. My understanding is that, for calibration, I need to call "bn_calibration_init" on all USBatchNorm2d objects, and then run a forward pass with a big batch_size, e.g., 1024.

4- I did calibration like I explained in step 3, but it did not give me good results. Interestingly, the above strategy does not work even with the uniform slimming, i.e., when all layers have the same with_multiplier. Can you please let me know where I got this wrong?

I would really appreciate your time if you can help me with this.

Mohammad

**Comment by JiahuiYu** [comment]
@MohammadSamragh Hi Mohammad, thanks for your interest in our work.

I guess you are loading our pre-trained US-Net models. However, the US-Net models we released are trained with uniform width_multipliers (which may explain your observation). To achieve the goal of [L1: 0.3, L2: 0.4, L3: 0.7, ...], please try to train a non-uniform model first as pre-trained weights.

**Comment by MohammadSamragh** [comment]
Thanks for the prompt response! I suspected that I should've trained the non-uniform model first. For that reason, I tried calibrating the pre-trained model with **uniform widths** (I calibrate again with one of the widths that were used during training), but that also gives me 0% accuracy. Can you read my step 3 (in the above description) and let me know if I am doing something wrong?

Many thanks!

**Comment by JiahuiYu** [comment]
@MohammadSamragh Hi I am a little bit confused on your settings. Here are my understandings. Correct me if I am wrong.

In your 2nd point:
Case 1. You calibrated with [L1: 0.3, L2: 0.3, L3: 0.3, ...] using our code, and then test **also** with [L1: 0.3, L2: 0.3, L3: 0.3, ...], the accuracy is OK.
Case 2. You calibrated with [L1: 0.3, L2: 0.4, L3: 0.7, ...] using our code, and then test **also** with [L1: 0.3, L2: 0.4, L3: 0.7, ...], the accuracy is very bad (because our released models are all trained with uniform widths).

If the above two cases are what you are doing, then I cannot understand your 3rd point:

> The above problem was inevitable since the batchnorm layers are only calibrated for uniform width_multipliers.

In case 2, you are already calibrating with non-uniform width_multipliers, right?

OR

if you are not doing Case 2, instead you are doing

Case 3: You calibrated with [L1: 0.3, L2: 0.3, L3: 0.3, ...] using our code, and then test **instead** with [L1: 0.3, L2: 0.4, L3: 0.7, ...], the accuracy is very bad.

**Comment by MohammadSamragh** [comment]
Sorry I think I confused you with my second step. There I just test. I did not apply any calibration.

In case 3, I am doing something else:

NonUniform Test: Calibrate with  [L1: 0.3, L2: 0.4, L3: 0.7, ...],test it again with  [L1: 0.3, L2: 0.4, L3: 0.7, ...], the accuracy is very bad. (this is expected because the model was not originally trained in this setting)

Re-Calibrate pre-trained model with [L1: 0.3, L2: 0.3, L3: 0.3, ...], test it again with [L1: 0.3, L2: 0.3, L3: 0.3, ...], the accuracy drops from 72% to 0%. (not extected, because re-calibrating a pre-trained model should not hurt accuracy)

I really appreciate your time! Hope I could explain the question well.

**Comment by JiahuiYu** [comment]
@MohammadSamragh I see. That simply means on your side, the calibration is not working? On my side, I tried before I release the code that if I set width multiplier to 0.3 uniformly, calibrate the BN statistics, and then test it - the accuracy is good. In fact, that's exactly how the released statistics of 0.3 are computed.

I'd recommend carefully check what you have modified. Please let me know if it still does not work on your side. Thanks!

**Comment by changlin31** [comment]
Thanks for the issue.
